### PR TITLE
FTL coordinate disk command for admins: ftldisk

### PR DESCRIPTION
**Pull request by IProduceWidgets**
## About the PR

Expand NanoTrasen's dominion over all ~~time and~~ space once more.

## Why / Balance

Let the admins play with shuttles across multiple maps with less micromanagement.

## Technical details

Takes a list of EntIDs, checks if theyre a map, if they aren't, finds the map they are on, then makes sure the map is setup as a destination, and finally spawns a disk setup for the map.

## Media

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:ADMIN:
- add: new command "ftldisk [EntID]" to make coordinate disks for shuttles to go to other maps.

### CODE REVIEW

**Comment by IProduceWidgets** [comment]
Okay i cooked my master branch a bit apparently, but its fiiiiine, it'll squash.

**Comment by 0x6273** [comment]
`IConsoleCommand` is deprecated. Use toolshed for new commands.

**Comment by IProduceWidgets** [comment]
> `IConsoleCommand` is deprecated. Use toolshed for new commands.

I thought this was backtracked?

**Comment by moonheart08** [comment]
> > `IConsoleCommand` is deprecated. Use toolshed for new commands.
> 
> I thought this was backtracked?

Not serverside. Use toolshed for this.

**Comment by TheDoctor1977** [comment]
Is this still being worked on?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
- Needs updating to code standards for commands.
- Needs a completion helper.

**Comment by IProduceWidgets** [comment]
> * Needs a completion helper.
image

> * Needs updating to code standards for commands.
I've looked at this for a couple weeks, and I have no idea what you mean lmao.

**Comment by IProduceWidgets** [comment]
I swapped from "Map ID" -> "Map netID" since we have two schemas

**Comment by 0x6273** [comment]
> > * Needs updating to code standards for commands.
> 
> I've looked at this for a couple weeks, and I have no idea what you mean lmao.

For new commands the standard is to use toolshed, unless there's a specific reason you need to use the old command system.

Do you need any help with porting this to toolshed?

**Comment by IProduceWidgets** [comment]
> > > * Needs updating to code standards for commands.
> > 
> > 
> > I've looked at this for a couple weeks, and I have no idea what you mean lmao.
> 
> For new commands the standard is to use toolshed, unless there's a specific reason you need to use the old command system.
> 
> Do you need any help with porting this to toolshed?

If you want to be my guest. I looked into it a bit, and it seems it will need engine stuff I do not want to do, and every time I ask about it I get conflicting direction.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs]: That was surprisingly painless to switch to.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs]: You're right it shouldn't ever get hit since it _should_ return telling you there's not enough arguments. But just in case I didn't think of something like trying to fly to nullspace I went ahead and did both. Either way it shouldn't die if you somehow manage to do that on live.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:62]: hmm. Does AssertNotNull handle null vars when in release? It shouldn't matter here, but I thought it didn't. Although Visual Studio sure seems to think it does.🥴

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:85]: I think this one is fine? C# just doesn't understand that the if handles the null implicitly. There might be another way to do that which is safer than just telling it to ignore it though. I can add another debug assert it nothing else.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:92]: Unfortunately we're subtly changing from checking any entity in the previous if, to checking what should only be a map in this one. Trying to squeeze it into the previous one would I think mean giving up on handling non-map entities by finding their map.

Or to put it another way, it will need to check whether dest is a map twice regardless since it changes what dest is (from any random entity to hopefully a map entity). Not nesting the if just handles scoping without defining it beforehand.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:68]: ''

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:92]: Well, hold on maybe I can put all three of the later checks in...

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs]: I couldn't find any other explicit casts this way in the repo so I asked in Howdoicode and got the suggestion to use !.Value and a comment.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs]: hmmm.... I suppose TryGetMap would probably return the map anyway if you passed it a map entity. I'm not sure that saves us anything though since (ideally) this whole blocks wont be running 95% of the time anyway since it tries to autocomplete a map entity in the helper.

Its probably just a smidge better to keep it the way it is though since doing that would mean running the TryGetComponent<TransformComponent> and TryGetMap calls 100% of the time when they're unlikely to be needed.

**Comment by IProduceWidgets** [comment]
There are 3 issues you may encounter with the command, but they aren't issues with the command so much as they are issues with maps which 'mattered until now.
I'm going to try to fix them in separate PRs, but I'll list them here for reference.

1. The FTL map can't be FTL'ed to, probably due to the FTL console/system thinking that's a dumb idea (it kinda is, but idk maybe you want to do warp speed pirates or something).
2. Centcom has the FTLDestinationComponent, but its not setup properly and needs to have the component's Enabled and RequireCoordinatesDisk fields set to true. Its also got a generic "Map Entity" name for the map.
3. Gateway Destination maps need to spawn and name an FTL point along with the gateway entity. They're surprisingly hard to get to and require a lot of fiddling with F3 or the lsmap command to tp to them.

But don't worry! This just works tm for new maps you create and initialize or for your admin test area. So you could for example use this to have two stations on different maps and let the players ferry themselves between them via FTL.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:24]: Don't include capital letters in commands, and make the command more plain and descriptive, i.e. something like `ftlcreatedisk`.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:173]: Double-check that this completion works with the current master; it seemed to me like it stopped working after merging master into my test PR branch.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:90]: Leave the swearing to code comments, not things that get written out in the console. 😉

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs:152]: I would prefer if the disk ended up the entity's hand if they have one. Otherwise an admin might end up spawning the disk on the floor unintentionally.

I think you can look at stuff like the Uplink for the correct function for that.

**Review by SlamBamActionman — CHANGES_REQUESTED** [review]
I've looked through this PR now and it seems to work as intended. Sometimes it seems like I need to add FTLdestination component to the target map or make edits to the beacons, but I feel that's something which could end up in some documentation rather than the command itself.

Don't worry about FTLing to the FTL map, as that should probably be off-limits. The parallax scrolling is caused by ships moving through it anyways, so you can't have a constant scroll just by being on the map. 

Let me know once you've addressed the requested changes!

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs]: I can make it try. The command makes an arbitrary number of disks so ymmv.

File: Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs
Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs]: The completion works, someone just messed up the helper in engine so it returns empty strings.
https://github.com/space-wizards/RobustToolbox/commit/26b09283f3e8ad60842a1564743f4da491c17c7e#diff-fcc1817dc485c3ac2fc001b91638163a3c8e4b37a75ce4a474840e83028ac44eL147

Comment (IProduceWidgets) [reply Content.Server/Shuttles/Commands/FTLDiskBurnerCommand.cs]: Made an engine PR to fix it, but I wouldn't hold this up on account of that. Its just a help text.
https://github.com/space-wizards/RobustToolbox/pull/5495

**Review by SlamBamActionman — APPROVED** [review]
Looks good to me!

File: Resources/Locale/en-US/shuttles/commands.ftl
Comment (SlamBamActionman) [inline Resources/Locale/en-US/shuttles/commands.ftl:2]: `coodrinates` > `coordinates`

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:38]: Requires localization (see https://github.com/space-wizards/space-station-14/blob/2a22964e512eecb06d8b7c96e92c1b1dc2ffcc75/Resources/Locale/en-US/shell.ftl for some generic ones)

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:46]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:52]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:82]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:88]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:99]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:104]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:114]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:133]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:145]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:152]: Localization

File: Content.Server/Shuttles/Commands/FTLDiskCommand.cs
Comment (SlamBamActionman) [inline Content.Server/Shuttles/Commands/FTLDiskCommand.cs:205]: Localization

**Review by SlamBamActionman — CHANGES_REQUESTED** [review]
Slart pointed out that the command should have localization to the error messages, so that has to be fixed first. Testing works well though 👍

**Comment by IProduceWidgets** [comment]
Localized

**Review by SlamBamActionman — APPROVED** [review]

**Pull request merged**
